api/v1alpha1/clusterconfig_types.go

```diff
@@ -44,81 +41,125 @@ var (
+	return dockerClusterConfigVariableSchema
+}
+
+// DockerClusterConfigSpec defines the desired state of DockerClusterConfig.
+type DockerClusterConfigSpec struct {
 	// +optional
 	Docker *DockerSpec `json:"docker,omitempty"`
+
+	GenericClusterConfigSpec `json:",inline"`
+
+	// +optional
+	ControlPlane *DockerNodeConfigSpec `json:"controlPlane,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// NutanixClusterConfig is the Schema for the nutanixclusterconfigs API.
+type NutanixClusterConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec NutanixClusterConfigSpec `json:"spec,omitempty"`
+}
+
+func (s NutanixClusterConfig) VariableSchema() clusterv1.VariableSchema { //nolint:gocritic,lll // Passed by value for no potential side-effect.
+	return nutanixClusterConfigVariableSchema
+}
+
+// NutanixClusterConfigSpec defines the desired state of NutanixClusterConfig.
+type NutanixClusterConfigSpec struct {
 	// +optional
 	Nutanix *NutanixSpec `json:"nutanix,omitempty"`
 
-	GenericClusterConfig `json:",inline"`
+	GenericClusterConfigSpec `json:",inline"`
 
 	// +optional
-	ControlPlane *NodeConfigSpec `json:"controlPlane,omitempty"`
+	ControlPlane *NutanixNodeConfigSpec `json:"controlPlane,omitempty"`
 }
 
-func (s ClusterConfigSpec) VariableSchema() clusterv1.VariableSchema { //nolint:gocritic,lll // Passed by value for no potential side-effect.
-	clusterConfigProps := GenericClusterConfig{}.VariableSchema()
-	switch {
-	case s.AWS != nil:
-		maps.Copy(
-			clusterConfigProps.OpenAPIV3Schema.Properties,
-			map[string]clusterv1.JSONSchemaProps{
-				AWSVariableName: s.AWS.VariableSchema().OpenAPIV3Schema,
-				"controlPlane":  s.ControlPlane.VariableSchema().OpenAPIV3Schema,
-			},
-		)
-	case s.Docker != nil:
-		maps.Copy(
-			clusterConfigProps.OpenAPIV3Schema.Properties,
-			map[string]clusterv1.JSONSchemaProps{
-				"docker": DockerSpec{}.VariableSchema().OpenAPIV3Schema,
-				"controlPlane": NodeConfigSpec{
-					Docker: &DockerNodeSpec{},
-				}.VariableSchema().OpenAPIV3Schema,
-			},
-		)
-	case s.Nutanix != nil:
-		maps.Copy(
-			clusterConfigProps.OpenAPIV3Schema.Properties,
-			map[string]clusterv1.JSONSchemaProps{
-				NutanixVariableName: NutanixSpec{}.VariableSchema().OpenAPIV3Schema,
-				"controlPlane": NodeConfigSpec{
-					Nutanix: &NutanixNodeSpec{},
-				}.VariableSchema().OpenAPIV3Schema,
-			},
-		)
-	}
+// +kubebuilder:object:root=true
 
-	return clusterConfigProps
+// GenericClusterConfig is the Schema for the clusterconfigs API.
+type GenericClusterConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec GenericClusterConfigSpec `json:"spec,omitempty"`
 }
 
-func NewAWSClusterConfigSpec() *ClusterConfigSpec {
-	return &ClusterConfigSpec{
-		AWS: &AWSSpec{},
-		ControlPlane: &NodeConfigSpec{
-			AWS: NewAWSControlPlaneNodeSpec(),
```

Comment:
We can do this, but we will need to duplicate things a little in the CP and worker node specs. Personally I think this is OK as it becomes more explicit API definition rather than a functional difference.

Reply:
If you look at [`e173208` (#537)](https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/537/commits/e173208bce30db54c8890f0fb7c8852aef16de47#diff-76fd5c53b3ae822e358b8641fd36454e158410ccbf318278343b5c4f513cc954) you can see how we achieve the different defaults now, by using different type definitions...

Reply:
Ahh yep, I see it! (putting this here because links to commits break after force pushes)

```
	// +kubebuilder:default=m5.xlarge
	// +optional
	InstanceType string `json:"instanceType,omitempty"`
```

```
	// The AWS instance type to use for the cluster Machines.
	// +kubebuilder:default=m5.2xlarge
	// +optional
	InstanceType string `json:"instanceType,omitempty"`
```

Reply:
Please remove these unused vars
```
	AWSControlPlaneInstanceType = "m5.xlarge"
	AWSWorkerInstanceType       = "m5.2xlarge"

	AWSControlPlaneInstanceProfile = "control-plane.cluster-api-provider-aws.sigs.k8s.io"
	AWSWorkerInstanceProfile       = "nodes.cluster-api-provider-aws.sigs.k8s.io"
```

Reply:
Done!